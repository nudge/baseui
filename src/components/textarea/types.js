/*
MIT License

Copyright (c) 2018 Uber Technologies, Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/
// @flow
import * as React from 'react';
import type {BaseInputPropsT, StateReducerT, StateT} from '../input/types';
import {SIZE} from '../input/constants';
import type {OverrideT} from '../../helpers/overrides';
import type {ThemeT} from '../../styles/types';

type SyntheticTextareaEvent = SyntheticEvent<HTMLElement>;

export type SizeT = $Keys<typeof SIZE>;

export type SharedStylePropsT = {
  $theme: ThemeT,
};

export type TextareaComponentsT = {
  InputContainer?: OverrideT<SharedStylePropsT>,
  Input?: OverrideT<SharedStylePropsT>,
};

export type TextareaPropsT = {
  ...BaseInputPropsT,
  components?: TextareaComponentsT,
  rows?: number,
};

export type StatefulContainerPropsT = {
  children: (props: *) => React.Node,
  initialState?: StateT,
  stateReducer?: StateReducerT,
  onChange?: (e: SyntheticTextareaEvent) => void,
};

type OmitPropsT = {
  children: (props: *) => React.Node,
};

type FullStPropsT = TextareaPropsT & StatefulContainerPropsT;

export type StatefulTextareaPropsT = $Diff<FullStPropsT, OmitPropsT>;
