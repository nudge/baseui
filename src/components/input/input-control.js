// @flow
import Input from './input';
import {asFormControl} from '../form-control';
// in the case we want to export both a FormControlHOC and a FormControl component
// we'll need to export additional InputControl `export default asFormControl(Input);`
// along with the Input component `export default Input;`
export default asFormControl(Input);
