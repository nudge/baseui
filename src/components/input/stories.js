/*
MIT License

Copyright (c) 2018 Uber Technologies, Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/
// @flow
/*global module */
import * as React from 'react';
import {storiesOf} from '@storybook/react';
import {withStyle} from 'styletron-react';
import {styled} from '../../styles';
import {withReadme} from 'storybook-readme';
//$FlowFixMe
import InputReadme from '../../../rfcs/input-component.md';
import {
  Input as ControlledInput,
  StatefulInput as Input,
  StyledInputContainer,
  SIZE,
} from './index';

const Button = styled('button', ({$theme}) => {
  return {
    ...$theme.typography.font300,
    display: 'block',
    paddingTop: '10px',
    paddingRight: '12px',
    paddingBottom: '10px',
    paddingLeft: '12px',
    marginTop: '8px',
    width: '100%',
    borderRadius: $theme.sizing.scale100,
    borderWidth: 'none',
  };
});

const InputIcon = styled('span', props => {
  return {
    width: '16px',
    display: 'flex',
    alignItems: 'center',
    padding:
      props.$position === 'left'
        ? '0 0 0 12px'
        : props.$position === 'right' ? '0 12px 0 0' : '0',
    ':before': {
      content: '""',
      display: 'inline-block',
      boxSizing: 'border-box',
      verticalAlign: 'middle',
      width: '16px',
      height: '16px',
      borderRadius: '50%',
      backgroundColor: props.$isFocused
        ? props.$theme.colors.primary
        : '#999999',
    },
  };
});

const RootWithStyle = withStyle(StyledInputContainer, props => {
  const {$disabled, $error, $isFocused, $theme: {colors, sizing}} = props;
  return {
    borderColor: $disabled
      ? colors.borderAlt
      : $error
        ? colors.borderError
        : $isFocused ? 'darkseagreen' : colors.border,
    boxShadow: `0 0 ${sizing.scale100} ${
      $disabled
        ? 'transparent'
        : $error
          ? colors.shadowError
          : $isFocused ? 'lightseagreen' : 'transparent'
    }`,
  };
});

const TextHighlight = styled('span', ({$theme}) => {
  return {color: $theme.colors.primary};
});

storiesOf('Input', module)
  .addDecorator(withReadme(InputReadme))
  // .add('Passing dynamic props', () => {
  //   return <MyComponent />;
  // })
  .add('Controlled and uncontrolled input', () => {
    return (
      <React.Fragment>
        <ControlledInput placeholder="Controlled input" />
        <br />
        <Input placeholder="Uncontrolled (stateful) input" />
      </React.Fragment>
    );
  })
  .add('Input size', () => {
    return (
      <React.Fragment>
        <Input placeholder="Default input" />
        <br />
        <Input size={SIZE.compact} placeholder="Compact input" />
      </React.Fragment>
    );
  })
  .add('Input state', () => {
    return (
      <React.Fragment>
        <Input size={SIZE.compact} placeholder="Default input" />
        <br />
        <Input
          initialState={{value: 'uber'}}
          autoFocus
          startEnhancer="@"
          endEnhancer=".com"
          size={SIZE.compact}
          placeholder="Initially focused input"
        />
        <br />
        <Input
          size={SIZE.compact}
          error
          placeholder="Input in an error state"
        />
        <br />
        <Input
          startEnhancer="@"
          endEnhancer=".com"
          size={SIZE.compact}
          error
          placeholder="Input with enhancers in an error state"
        />
        <br />
        <Input size={SIZE.compact} disabled placeholder="Disabled input" />
        <br />
        <Input
          startEnhancer="@"
          endEnhancer=".com"
          size={SIZE.compact}
          disabled
          placeholder="Disabled input with enhancers"
        />
      </React.Fragment>
    );
  })
  .add('Input enhancers', () => {
    return (
      <React.Fragment>
        <Input startEnhancer="@" placeholder="Input with a startEnhancer" />
        <br />
        <Input endEnhancer=".00" placeholder="Input with an endEnhancer" />
        <br />
        <Input
          startEnhancer="@"
          endEnhancer=".00"
          placeholder="Input with start and end enhancers"
        />
        <br />
        <Input
          startEnhancer={<InputIcon />}
          endEnhancer={<InputIcon />}
          placeholder="Input with element type enhancers"
        />
        <br />
        <Input
          startEnhancer={({$isFocused}) => {
            return $isFocused ? <TextHighlight>@</TextHighlight> : '@';
          }}
          endEnhancer={({$isFocused}) => {
            return $isFocused ? <TextHighlight>.00</TextHighlight> : '.00';
          }}
          placeholder="Input with function type enhancers. Shared props are passed into an enhancer func"
        />
      </React.Fragment>
    );
  })
  .add('Input with Before and After', () => {
    return (
      <React.Fragment>
        <Input
          overrides={{
            Before: function Before(props) {
              return <InputIcon {...props} $position="left" />;
            },
          }}
          placeholder="Input with a Before component"
        />
        <br />
        <Input
          overrides={{
            After: function After(props) {
              return <InputIcon {...props} $position="right" />;
            },
          }}
          placeholder="Input with an After component"
        />
      </React.Fragment>
    );
  })
  .add(
    'Input with a with custom component override, style overrides and extra props passed',
    () => {
      return (
        <React.Fragment>
          <Input
            overrides={{InputContainer: {component: RootWithStyle}}}
            autoFocus
            placeholder="Input with a custom InputContainer override"
          />
          <br />
          <Input
            overrides={{
              Input: {
                props: {'data-test': 'test'},
                style: {color: '#00F'},
              },
            }}
            initialState={{
              value:
                "With a 'data-test' attrs passes to the input and text style override",
            }}
            placeholder="With a 'data-test' attrs passes to the input and text style override"
          />
        </React.Fragment>
      );
    },
  )
  .add('Input with a ref', () => {
    const inputRef: {current: ?React.ElementRef<'input'>} = React.createRef();
    return (
      <React.Fragment>
        <Input inputRef={inputRef} placeholder="With input ref" />
        <Button
          onClick={() => {
            inputRef.current && inputRef.current.focus();
          }}
        >
          Click here to focus input
        </Button>
      </React.Fragment>
    );
  });
