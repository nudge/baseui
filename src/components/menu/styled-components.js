/*
MIT License

Copyright (c) 2018 Uber Technologies, Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/
// @flow
import {styled} from '../../styles';

import type {ThemeT} from '../../styles';

type StyledPropsT = {
  $theme: ThemeT,
};

type StyledListItemPropsT = {
  $isHighlighted: boolean,
} & StyledPropsT;

export const List = styled('ul', ({$theme}: StyledPropsT) => ({
  position: 'relative',
  display: 'flex',
  flexDirection: 'column',
  margin: 0,
  background: $theme.colors.white,
  borderRadius: $theme.borders.radius300,
  boxShadow: $theme.lighting.shadow600,
  paddingTop: $theme.sizing.scale200,
  paddingBottom: $theme.sizing.scale200,
  paddingLeft: $theme.sizing.scale400,
  paddingRight: $theme.sizing.scale400,
}));

export const ListItem = styled(
  'li',
  ({$theme, $isHighlighted}: StyledListItemPropsT) => ({
    ...$theme.typography.font400,
    position: 'relative',
    display: 'block',
    color: $isHighlighted ? $theme.colors.primary : $theme.colors.black,
    margin: 0,
    cursor: 'pointer',
    transitionProperty: 'color',
    transitionDuration: $theme.animation.timing100,
    transitionTimingFunction: $theme.animation.easeOutCurve,
    ':hover': {
      color: $theme.colors.primary400,
    },
  }),
);
