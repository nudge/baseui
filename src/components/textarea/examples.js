/*
MIT License

Copyright (c) 2018 Uber Technologies, Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/
// @flow
import * as React from 'react';
import {withStyle} from 'styletron-react';
import {styled} from '../../styles';
import {withProps} from '../../helpers';
import {
  StatefulTextarea as Textarea,
  Textarea as ControlledTextarea,
  StyledTextarea,
  SIZE,
} from './index';

const Button = styled('button', ({$theme}) => {
  return {
    ...$theme.typography.font300,
    display: 'block',
    paddingTop: '10px',
    paddingRight: '12px',
    paddingBottom: '10px',
    paddingLeft: '12px',
    marginTop: '8px',
    width: '100%',
    borderRadius: $theme.sizing.scale100,
    borderWidth: 'none',
  };
});

const TextareaWithStyle = withStyle(StyledTextarea, props => {
  const {$disabled, $error, $isFocused, $theme: {colors, sizing}} = props;
  return {
    borderColor: $disabled
      ? colors.borderAlt
      : $error
        ? colors.borderError
        : $isFocused ? 'darkseagreen' : colors.border,
    boxShadow: `0 0 ${sizing.scale100} ${
      $disabled
        ? 'transparent'
        : $error
          ? colors.shadowError
          : $isFocused ? 'lightseagreen' : 'transparent'
    }`,
  };
});

const TextareaWithProps = withProps(StyledTextarea, {
  'data-test': 'test',
});

export const tests = {
  SIMPLE_EXAMPLE: 'Controlled and uncontrolled textarea',
  SIZE_EXAMPLE: 'Textarea size',
  STATE_EXAMPLE: 'Textarea state',
  OVERRIDES_EXAMPLE: 'Textarea with style overrides and extra props passed',
  REF_EXAMPLE: 'Textarea with a ref',
};

export default {
  [tests.SIMPLE_EXAMPLE]: function Story1() {
    return (
      <React.Fragment>
        <ControlledTextarea placeholder="Controlled textarea" />
        <br />
        <Textarea
          placeholder="Uncontrolled textarea"
          initialState={{value: 'initial value'}}
        />
      </React.Fragment>
    );
  },
  [tests.SIMPLE_EXAMPLE]: function Story2() {
    return (
      <React.Fragment>
        <Textarea placeholder="Default textarea" />
        <br />
        <Textarea size={SIZE.compact} placeholder="Compact textarea" />
      </React.Fragment>
    );
  },
  [tests.STATE_EXAMPLE]: function Story3() {
    return (
      <React.Fragment>
        <Textarea size={SIZE.compact} placeholder="Placeholder" />
        <br />
        <Textarea
          initialState={{value: 'uber'}}
          autoFocus
          size={SIZE.compact}
          placeholder="Placeholder"
        />
        <br />
        <Textarea size={SIZE.compact} error={true} placeholder="Placeholder" />
        <br />
        <Textarea
          size={SIZE.compact}
          disabled
          placeholder="Disabled textarea"
        />
      </React.Fragment>
    );
  },
  [tests.OVERRIDES_EXAMPLE]: function Story4() {
    return (
      <React.Fragment>
        <Textarea
          autoFocus
          overrides={{Input: TextareaWithStyle}}
          placeholder="With style overrides on the textarea element"
        />
        <br />
        <Textarea
          label="Input with extra props"
          overrides={{
            Input: TextareaWithProps,
          }}
          placeholder="With a 'data-test' attrs passes to the input element"
        />
      </React.Fragment>
    );
  },
  [tests.REF_EXAMPLE]: function Story5() {
    const inputRef: {current: ?React.ElementRef<'input'>} = React.createRef();
    return (
      <React.Fragment>
        <Textarea inputRef={inputRef} placeholder="With textarea input ref" />
        <Button
          onClick={() => {
            inputRef.current && inputRef.current.focus();
          }}
        >
          Click here to focus input
        </Button>
      </React.Fragment>
    );
  },
};
