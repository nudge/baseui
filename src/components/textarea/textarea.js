/*
MIT License

Copyright (c) 2018 Uber Technologies, Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/
// @flow
import * as React from 'react';
import type {TextareaPropsT} from './types';
import {mergeOverrides} from '../../helpers/overrides';
import getBuiId from '../../utils/get-bui-id';
import {BaseInput, SIZE, CUSTOM_INPUT_TYPE} from '../input';
import {
  Textarea as StyledTextarea,
  TextareaContainer as StyledTextareaContainer,
} from './styled-components';

class Textarea extends React.Component<TextareaPropsT> {
  static defaultProps = {
    autoFocus: false,
    disabled: false,
    error: false,
    id: getBuiId(),
    inputRef: React.createRef(),
    name: '',
    onBlur: () => {},
    onChange: () => {},
    onFocus: () => {},
    overrides: {},
    placeholder: '',
    required: false,
    rows: 3,
    size: SIZE.default,
    value: '',
  };

  render() {
    const overrides = mergeOverrides(
      {
        Input: StyledTextarea,
        InputContainer: StyledTextareaContainer,
      },
      this.props.overrides,
    );
    return (
      <BaseInput
        {...this.props}
        type={CUSTOM_INPUT_TYPE.textarea}
        overrides={overrides}
      />
    );
  }
}

export default Textarea;
