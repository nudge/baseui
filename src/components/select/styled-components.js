// @flow
import {styled} from '../../styles';
import {ICON, TYPE} from './constants';

import {SIZE} from '../input';
import {
  getInputStyles,
  getInputContainerStyles,
} from '../input/styled-components';

import {getListStyles, getListItemStyles} from '../menu/styled-components';

export const Root = styled('div', props => {
  return {
    position: 'relative',
  };
});

export const Input = styled('input', props => {
  return {
    ...getInputStyles({...props, $size: SIZE.default}),
    cursor: 'pointer',
    width: 'auto',
    flexGrow: '1',
    ':hover': {
      cursor: 'pointer',
    },
  };
});

export const InputContainer = styled('div', props => {
  const {$theme} = props;
  const {
    colors: {primary400},
  } = $theme;
  return {
    ...getInputContainerStyles({...props, $size: SIZE.default}),
    flexWrap: 'wrap',
    padding: '5px',
    alignItems: 'center',
    position: 'relative',
    backgroundColor: 'white',
    borderColor: primary400,
  };
});

export const Tag = styled('span', props => {
  const {$theme, $multiple} = props;
  const {
    colors: {primary400, mono400},
    sizing: {scale800},
  } = $theme;
  return $multiple
    ? {
        fontWeight: 'bold',
        display: 'flex',
        alignItems: 'center',
        padding: '5px',
        margin: '5px',
        borderWidth: '1px',
        borderColor: primary400,
        color: primary400,
        borderRadius: '7px',
        lineHeight: scale800,
        backgroundColor: mono400,
      }
    : {
        ...getInputStyles({...props, $size: SIZE.default, $disabled: true}),
        cursor: 'pointer',
        width: 'auto',
        flexGrow: '1',
        ':hover': {
          cursor: 'pointer',
        },
      };
});

export const SearchIcon = styled('img', props => {
  const {$theme} = props;
  const {
    sizing: {scale300, scale400, scale500},
  } = $theme;
  switch (props.$type) {
    case ICON.CLEAR_ALL:
      return {
        marginLeft: 'auto',
        position: 'absolute',
        right: scale400,
      };
    case ICON.SELECT:
      return {
        marginRight: scale500,
      };
    case ICON.SELECTED:
      return {
        paddingRight: scale300,
      };
    case ICON.CLEAR_TAG:
    case ICON.LOOP:
    default:
      return {};
  }
});

export const DropDown = styled('ul', props => {
  const {$theme, $isOpen, $type, $rows} = props;
  const {
    sizing: {scale300, scale600, scale1000},
  } = $theme;
  return {
    ...getListStyles(props),
    height: $rows ? parseInt(scale600) * $rows + 'px' : null,
    overflowY: $rows ? 'scroll' : null,
    top: $type === TYPE.SELECT ? scale1000 : null,
    display: !$isOpen ? 'none' : null,
    width: '96%',
    position: 'absolute',
    padding: scale600,
    listStyle: 'none',
    borderRadius: scale300,
    boxShadow: $theme.lighting.shadow600,
  };
});

export const DropDownItem = styled('li', props => {
  const {$theme} = props;
  const {
    sizing: {scale600},
  } = $theme;
  return {
    ...getListItemStyles(props),
    lineHeight: scale600,
  };
});

export const Option = styled('div', props => {
  const {$selected, disabled, $theme} = props;
  const {
    colors: {mono700, primary400},
    sizing: {scale300},
  } = $theme;
  const padding = $selected
    ? {
        paddingTop: scale300,
        paddingBottom: scale300,
        paddingRight: '0px',
        paddingLeft: '0px',
      }
    : {
        paddingTop: scale300,
        paddingBottom: scale300,
        paddingRight: '18px',
        paddingLeft: '18px',
      };
  return {
    ':hover': {
      cursor: disabled ? 'not-allowed' : 'text',
    },
    color: disabled ? mono700 : $selected ? primary400 : null,
    ...padding,
  };
});
