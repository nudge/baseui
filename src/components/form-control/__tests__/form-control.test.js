/*
MIT License

Copyright (c) 2018 Uber Technologies, Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/
// @flow
import React from 'react';
import {mount} from 'enzyme';
import FormControl from '../form-control';
import {Label, Caption} from '../styled-components';
import {Input} from '../../input';
import {Textarea} from '../../textarea';
import {Checkbox} from '../../checkbox';
import {RadioGroup, StyledRadio} from '../../radio';

describe('FormControl - Label and Caption for controls', () => {
  test('Renders label, caption, and error for the Input component', () => {
    const rendered = mount(
      <FormControl label="Label test" caption="Caption test">
        <Input />
      </FormControl>,
    );
    const label = rendered.find(Label).first();
    expect(label).toExist();
    expect(label).toHaveText('Label test');
    expect(label.props()).toMatchSnapshot('Label gets correct props');
    const caption = rendered.find(Caption).first();
    expect(caption).toExist();
    expect(caption).toHaveText('Caption test');
    expect(caption.props()).toMatchSnapshot('Caption gets correct props');

    rendered.setProps({
      error: 'Error test',
    });
    expect(caption).toHaveText('Error test');
    expect(caption.props()).toMatchSnapshot(
      'Caption gets correct props when error is passed',
    );
  });

  test('Accepts a node for label, caption, and error', () => {
    const rendered = mount(
      <FormControl
        label={<span>Label test</span>}
        caption={<span>Caption test</span>}
      >
        <Input />
      </FormControl>,
    );
    const label = rendered.find(Label).first();
    expect(label).toHaveText('Label test');
    const caption = rendered.find(Caption).first();
    expect(caption).toHaveText('Caption test');

    rendered.setProps({
      error: <span>Error test</span>,
    });
    expect(caption).toHaveText('Error test');
  });

  test('Accepts a function for label, caption, and error', () => {
    const label = jest.fn().mockReturnValue(<span>Label test</span>);
    const caption = jest.fn().mockReturnValue(<span>Caption test</span>);
    const error = jest.fn().mockReturnValue(<span>Error test</span>);
    const rendered = mount(
      <FormControl label={label} caption={caption}>
        <Input />
      </FormControl>,
    );
    const sharedProps = {
      $disabled: false,
      $error: false,
      $required: false,
      $size: 'default',
    };
    const labelRendered = rendered.find(Label).first();
    expect(label).toHaveBeenCalledWith(sharedProps);
    expect(labelRendered).toHaveText('Label test');
    const captionRendered = rendered.find(Caption).first();
    expect(caption).toHaveBeenCalledWith(sharedProps);
    expect(captionRendered).toHaveText('Caption test');
    expect(error).not.toHaveBeenCalled();

    rendered.setProps({
      error: error,
    });
    expect(error).toHaveBeenCalledWith({
      ...sharedProps,
      $error: error,
    });
    expect(captionRendered).toHaveText('Error test');
  });

  test('Renders caption if error is set on the control', () => {
    const rendered = mount(
      <FormControl label="Label test" caption="Caption test">
        <Input error />
      </FormControl>,
    );
    const caption = rendered.find(Caption).first();
    expect(caption).toHaveText('Caption test');
    expect(caption.props()).toMatchSnapshot(
      'Caption gets correct props if error set on the control',
    );
  });

  test('Passes correct props from the control to label and caption', () => {
    const rendered = mount(
      <FormControl label="Label test" caption="Caption test">
        <Input disabled required error />
      </FormControl>,
    );
    const label = rendered.find(Label).first();
    expect(label.props()).toMatchSnapshot(
      'Label gets correct props from the control',
    );
    const caption = rendered.find(Caption).first();
    expect(caption.props()).toMatchSnapshot(
      'Caption gets correct props from the control',
    );
    rendered.setProps({
      error: 'Error test',
    });
    expect(label.props()).toMatchSnapshot('Label gets correct error value');
    expect(caption.props()).toMatchSnapshot('Caption gets correct error value');
    expect(caption).toHaveText('Error test');
  });

  test('Renders label and caption for the Teaxtarea component', () => {
    const rendered = mount(
      <FormControl label="Label test" caption="Caption test">
        <Textarea required />
      </FormControl>,
    );
    const label = rendered.find(Label).first();
    expect(label).toExist();
    expect(label.props()).toMatchSnapshot('Label gets correct props');
    const caption = rendered.find(Caption).first();
    expect(caption).toExist();
    expect(caption.props()).toMatchSnapshot('Caption gets correct props');
  });
});

test('Renders label and caption for the Checkbox component', () => {
  const rendered = mount(
    <FormControl label="Label test" caption="Caption test">
      <Checkbox required />
    </FormControl>,
  );
  const label = rendered.find(Label).first();
  expect(label).toExist();
  expect(label.props()).toMatchSnapshot('Label gets correct props');
  const caption = rendered.find(Caption).first();
  expect(caption).toExist();
  expect(caption.props()).toMatchSnapshot('Caption gets correct props');
});

test('Renders label and caption for the RadioGroup component', () => {
  const rendered = mount(
    <FormControl label="Label test" caption="Caption test">
      <RadioGroup required>
        <StyledRadio value="1">First</StyledRadio>
        <StyledRadio value="2">Second</StyledRadio>
        <StyledRadio value="3">Third</StyledRadio>
      </RadioGroup>
    </FormControl>,
  );
  const label = rendered.find(Label).first();
  expect(label).toExist();
  expect(label.props()).toMatchSnapshot('Label gets correct props');
  const caption = rendered.find(Caption).first();
  expect(caption).toExist();
  expect(caption.props()).toMatchSnapshot('Caption gets correct props');
});
