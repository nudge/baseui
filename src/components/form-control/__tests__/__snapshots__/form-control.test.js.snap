// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormControl - Label and Caption for controls Passes correct props from the control to label and caption: Caption gets correct error value 1`] = `
Object {
  "$disabled": true,
  "$error": true,
  "$required": true,
  "$size": "default",
  "children": "Caption test",
}
`;

exports[`FormControl - Label and Caption for controls Passes correct props from the control to label and caption: Caption gets correct props from the control 1`] = `
Object {
  "$disabled": true,
  "$error": true,
  "$required": true,
  "$size": "default",
  "children": "Caption test",
}
`;

exports[`FormControl - Label and Caption for controls Passes correct props from the control to label and caption: Label gets correct error value 1`] = `
Object {
  "$disabled": true,
  "$error": true,
  "$required": true,
  "$size": "default",
  "children": "Label test",
  "htmlFor": "bui-mock-id",
}
`;

exports[`FormControl - Label and Caption for controls Passes correct props from the control to label and caption: Label gets correct props from the control 1`] = `
Object {
  "$disabled": true,
  "$error": true,
  "$required": true,
  "$size": "default",
  "children": "Label test",
  "htmlFor": "bui-mock-id",
}
`;

exports[`FormControl - Label and Caption for controls Renders caption if error is set on the control: Caption gets correct props if error set on the control 1`] = `
Object {
  "$disabled": false,
  "$error": true,
  "$required": false,
  "$size": "default",
  "children": "Caption test",
}
`;

exports[`FormControl - Label and Caption for controls Renders label and caption for the Teaxtarea component: Caption gets correct props 1`] = `
Object {
  "$disabled": false,
  "$error": false,
  "$required": true,
  "$size": "default",
  "children": "Caption test",
}
`;

exports[`FormControl - Label and Caption for controls Renders label and caption for the Teaxtarea component: Label gets correct props 1`] = `
Object {
  "$disabled": false,
  "$error": false,
  "$required": true,
  "$size": "default",
  "children": "Label test",
  "htmlFor": "bui-mock-id",
}
`;

exports[`FormControl - Label and Caption for controls Renders label, caption, and error for the Input component: Caption gets correct props 1`] = `
Object {
  "$disabled": false,
  "$error": false,
  "$required": false,
  "$size": "default",
  "children": "Caption test",
}
`;

exports[`FormControl - Label and Caption for controls Renders label, caption, and error for the Input component: Caption gets correct props when error is passed 1`] = `
Object {
  "$disabled": false,
  "$error": false,
  "$required": false,
  "$size": "default",
  "children": "Caption test",
}
`;

exports[`FormControl - Label and Caption for controls Renders label, caption, and error for the Input component: Label gets correct props 1`] = `
Object {
  "$disabled": false,
  "$error": false,
  "$required": false,
  "$size": "default",
  "children": "Label test",
  "htmlFor": "bui-mock-id",
}
`;

exports[`Renders label and caption for the Checkbox component: Caption gets correct props 1`] = `
Object {
  "$disabled": false,
  "$error": undefined,
  "$required": true,
  "children": "Caption test",
}
`;

exports[`Renders label and caption for the Checkbox component: Label gets correct props 1`] = `
Object {
  "$disabled": false,
  "$error": undefined,
  "$required": true,
  "children": "Label test",
  "htmlFor": undefined,
}
`;

exports[`Renders label and caption for the RadioGroup component: Caption gets correct props 1`] = `
Object {
  "$disabled": false,
  "$error": undefined,
  "$required": true,
  "children": "Caption test",
}
`;

exports[`Renders label and caption for the RadioGroup component: Label gets correct props 1`] = `
Object {
  "$disabled": false,
  "$error": undefined,
  "$required": true,
  "children": "Label test",
  "htmlFor": undefined,
}
`;
